cage.go

```diff
@@ -0,0 +1,235 @@
+//	cageClient, err := evClient.CageClient(cageURL, []evervault.PCRs{expectedPCRs})
+//	if err != nil {
+//		log.Fatal(err)
+//	}
+//
+//	payload, err := json.Marshal(fmt.Sprintf(`{"encrypted": "%s"}`, encrypted))
+//	if err != nil {
+//		log.Fatal(err)
+//	}
+//
+//	req, _ := http.NewRequest(http.MethodPost, fmt.Sprintf("https://%s/", cageURL), bytes.NewBuffer(payload))
+//	req.Close = true
+//	req.Header.Set("API-KEY", "<API_KEY>")
+//	req.Header.Set("Content-Type", "application/json; charset=UTF-8")
+//
+//	resp, err := cageClient.Do(req)
+func (c *Client) CageClient(cageHostname string, expectedPCRs []PCRs) (*http.Client, error) {
+	c.expectedPCRs = filterEmptyPCRs(expectedPCRs)
+	if len(c.expectedPCRs) == 0 {
+		return nil, ErrNoPCRs
+	}
+
+	caCertResponse, err := c.makeRequest(c.Config.EvervaultCagesCaURL, http.MethodGet, nil, "")
+	if err != nil {
+		return nil, err
+	}
+
+	cagesClient, err := c.cagesClient(cageHostname, caCertResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	return cagesClient, nil
+}
+
+func (c *Client) cagesClient(cageHostname string, caCert []byte) (*http.Client, error) {
+	transport, err := c.cagesTransport(cageHostname, caCert)
+	if err != nil {
+		return nil, err
+	}
+
+	return &http.Client{Transport: transport}, nil
+}
+
+func (c *Client) cagesTransport(cageHostname string, caCert []byte) (*http.Transport, error) {
+	rootCAs, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, fmt.Errorf("error getting system cert pool %w", err)
+	}
+
+	rootCAs.AppendCertsFromPEM(caCert)
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: false,
+		RootCAs:            rootCAs,
+		MinVersion:         tls.VersionTLS12,
+		ServerName:         cageHostname,
+	}
+
+	customDial := c.createDial(tlsConfig)
+
+	return &http.Transport{
+		DisableKeepAlives: true,
+		DialTLSContext:    customDial,
+	}, nil
+}
+
+// Attest that a given cert matches the expected PCRS.
+func attestCert(cert []byte, expectedPCRs []PCRs) (bool, error) {
+	// 1) extract the X509Certificate certificate from the bytes
+	certificate, err := x509.ParseCertificate(cert)
+	if err != nil {
+		return false, fmt.Errorf("unable to parse certificate %w", err)
+	}
+	// Extract the largest DNS name from the certificate
+	largestIndex := 0
+	for i := 1; i < len(certificate.DNSNames); i++ {
+		if len(certificate.DNSNames[i]) > len(certificate.DNSNames[largestIndex]) {
+			largestIndex = i
+		}
+	}
+
+	coseDNSValue := certificate.DNSNames[largestIndex]
+	// extract the COSE signature from the DNS name
+	coseSig := strings.Split(coseDNSValue, ".")[0]
+	// decode the hex encoded COSE signature
+	hexDecodedDNS, err := hex.DecodeString(coseSig)
+	if err != nil {
+		return false, fmt.Errorf("unable to decode certificate %w", err)
+	}
+
+	res, err := nitrite.Verify(hexDecodedDNS, nitrite.VerifyOptions{CurrentTime: time.Now()})
+	if err != nil {
+		return false, fmt.Errorf("unable to verify certificate %w", err)
+	}
+
+	verified := verifyPCRs(expectedPCRs, *res.Document)
```

Comment:
We also need a check to validate that the public key of the certificate is included in the documents user data

Reply:
That the attestation doc's userdata field matches the public key of the x509 given in the server hello ([this step in the rust crate](https://github.com/evervault/attestation-doc-validation/blob/main/attestation-doc-validation/src/lib.rs#L70-L72))